### PR TITLE
Summary memory

### PR DESCRIPTION
**Pull request by Torantulino**
**I'm submitting this as a draft for now as this implementation cuts out sections of the current implementation and crudely inserts itself into the code. I didn't have time tonight to integrate this elegantly - I can either pick that up tomorrow or leave it for you.**

### Background
The current memory implementation is very flawed, testing has revealed that running autogpt with no memory currently performs better than with memory.

This is a basic attempt at fixing that. This implementation may be superseeded by - or combined with - a more advanced memory implementation in the near future. 

This will provide a good baseline to test such a future implementation against - proving that we're moving in the right direction.

### Changes
Adds a rolling summary memory system that identifies the new messages which have been trimmed from the full message history, and combines them into a summary of past events. This is inserted at the top of the message history to provide a clear overview of actions carried out by the agent in the past.

### Documentation
Full docstrings, comments and typing.

### Test Plan
Tested this by running the same prompt with the current and new system for ~20 iterations.
The memory output by this system can be seen below:

image

Compared to the original system, which in the same situation dumped 2000 tokens of irrelevant raw message history in the context window, this contains vastly more information in a tenth of the tokens.

This needs further work. I've noticed that - in rare cases - llm can make up an intiial fake summary when it is empty and there's nothing to add. Prompt tweaks should fix this.

### PR Quality Checklist
- [~] My pull request is atomic and focuses on a single change.
- [~] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: autogpt/llm/chat.py
Comment (Torantulino) [inline autogpt/llm/chat.py:1]: Commented out current/old memory code as it needs to be removed from throughout the codebase.

File: autogpt/memory_management/summary_memory.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    with an index higher than prev_index that are absent from current_context, and the new index.
+
+    Args:
+        full_message_history (list): A list of dictionaries representing the full message history.
+        current_context (list): A list of dictionaries representing the current context.
+        prev_index (int): An integer representing the previous index.
+
+    Returns:
+        list: A list of dictionaries that are in full_message_history with an index higher than prev_index and absent from current_context.
+        int: The new index value for use in the next loop.
```
Comment (Torantulino) [inline autogpt/memory_management/summary_memory.py]:
```suggestion
```

File: autogpt/memory_management/summary_memory.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+prev_index = 0
+current_memory = ""
+
+
+def get_newly_trimmed_messages(
+    full_message_history: List[Dict[str, str]], current_context: List[Dict[str, str]]
+) -> Tuple[List[Dict[str, str]], int]:
+    """
+    This function returns a list of dictionaries contained in full_message_history
+    with an index higher than prev_index that are absent from current_context, and the new index.
```
Comment (Torantulino) [inline autogpt/memory_management/summary_memory.py]:
```suggestion
    with an index higher than prev_index that are absent from current_context.
```

**Comment by dschonholtz** [comment]
Some comments on this:

This PR completely removes all of the vectorized memory in favor of a summarized memory. I am guessing that does perform significantly better so I think some version of it should go in today. (Will be checking shortly)

The question is what we do with all of the dead code? 

Presumably, vectorized memory will perform better with some heavy tweaking at some point in the future. 
1. Do we leave in all of the vectorized memory infra that would not be used as this goes in?  My guess is yes.
2. Do we remove all of the support infra/tests there that aren't used? My guess is yes, if the relevant code is deleted and no otherwise.

Also, currently the memory right now is saved as some global variable, so I'm going to move that at least to be controlled by a class.

Looking at agent manager, those agents don't really have any memory associated with them and don't do any memory management of their own, so that should all be fine.

File: autogpt/memory_management/summary_memory.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+from typing import Dict, List, Tuple
+
+from autogpt.config import Config
+from autogpt.llm.llm_utils import create_chat_completion
+
+cfg = Config()
```
Comment (ntindle) [inline autogpt/memory_management/summary_memory.py]: nit: Pass this in from the agent if possible to help the agent encapsulation project. 
CC: @collijk

**Comment by Torantulino** [comment]
> 
    Do we leave in all of the vectorized memory infra that would not be used as this goes in? My guess is yes.
    Do we remove all of the support infra/tests there that aren't used? My guess is yes, if the relevant code is deleted and no otherwise.

Agreed. We will likely use that again in the near future - or at least portions of it - so we don't want to remove and re-add the same code.

For now let's get this tested vs the current implementation - if it's better then get it in for the next release while we work on a more advanced system.

File: autogpt/memory_management/summary_memory.py (hunk 3 above)
Comment (Torantulino) [reply autogpt/memory_management/summary_memory.py]: Doing this gave me a circular dependency - I'll have to check how this is being done later

**Comment by Torantulino** [comment]
Testing underway by @merwanehamadi.

**Comment by ntindle** [comment]
This works significantly better than master from my testing, but after 6 or 7 iterations, it loses its memory.

File: autogpt/memory_management/summary_memory.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        elif event["role"] == "user":
+            new_events.remove(event)
+
+    # This can happen at any point during execturion, not just the beginning
+    if len(new_events) == 0:
+        new_events = "Nothing new happened."
+
+    prompt = f'''Your task is to create a concise running summary of actions and information results in the provided text, focusing on key and potentially important information to remember.
+
+You will receive the current summary and the your latest actions. Combine them, adding relevant key information from the latest development in 1st person past tense and keeping the summary concise.
```
Comment (ntindle) [inline autogpt/memory_management/summary_memory.py]: I had a better time with:
```suggestion
You will receive the current summary and your latest actions. Combine them, adding relevant key information from the latest development in 1st person past tense and keeping the summary concise. Never fully forget something from your memory.
```

**Comment by merwanehamadi** [comment]
@Torantulino beat the memory challenge B level 2 without memory, master wasn't beating it

**Pull request closed without merging**